## 4.5 Data Consistency Checks

### 1. Importing libraries

### 2. Importing data

### 3. Data consistency checks from Exercise

### 4. Task

#### 1. Importing libraries

In [4]:
# Import libraries
import pandas as pd
import numpy as np
import os

#### 2. Importing data

In [6]:
path = r'C:\Users\Priya\OneDrive\Documents\CareerFoundry\Data Immersion\Achievement 4\Instacart Basket Analysis'

In [7]:
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)

In [8]:
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = False)

#### 3. Data consistency checks from Exercise

In [10]:
#Create a dataframe
df_test = pd.DataFrame()

In [11]:
#Create a mixed-type column
df_test['mix'] = ['a', 'b', 1, True]

In [12]:
#Viewing df_test
df_test.head()

mix
0     a
1     b
2     1
3  True

In [13]:
df_test['mix'] = df_test['mix'].astype('str')

In [14]:
#Checking for missing data
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [15]:
#Creating df_nan subset
df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [16]:
#Viewing df_nan
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [17]:
#Checking the shape of df_prods
df_prods.shape

(49693, 5)

In [18]:
#Creating df_prods_clean subset
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [19]:
#Checking the shape of df_prods_clean
df_prods_clean.shape

(49677, 5)

In [20]:
#Creating the df_dups subset to check for duplicates
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [21]:
#Viewing df_dups
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [22]:
#Checking the shape of df_prods_clean
df_prods_clean.shape

(49677, 5)

In [23]:
#Creating a subset without duplicates
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [24]:
#Viewing df_prods_clean_no_dups
df_prods_clean_no_dups

product_id                                       product_name  \
0               1                         Chocolate Sandwich Cookies   
1               2                                   All-Seasons Salt   
2               3               Robust Golden Unsweetened Oolong Tea   
3               4  Smart Ones Classic Favorites Mini Rigatoni Wit...   
4               5                          Green Chile Anytime Sauce   
...           ...                                                ...   
49688       49684          Vodka, Triple Distilled, Twist of Vanilla   
49689       49685                 En Croute Roast Hazelnut Cranberry   
49690       49686                                   Artisan Baguette   
49691       49687         Smartblend Healthy Metabolism Dry Cat Food   
49692       49688                             Fresh Foaming Cleanser   

       aisle_id  department_id  prices  
0            61             19     5.8  
1           104             13     9.3  
2            94              7     4.5  
3            38              1    10.5  
4             5             13     4.3  
...         ...            ...     ...  
49688       124              5     5.3  
49689        42              1     3.1  
49690       112              3     7.8  
49691        41              8     4.7  
49692        73             11    13.5  

[49672 rows x 5 columns]

In [25]:
#Checking the shape of df_prods_clean_no_dups
df_prods_clean_no_dups.shape

(49672, 5)

In [26]:
#Checking statistics for df_prods_clean_no_dups
df_prods_clean_no_dups.describe()

product_id      aisle_id  department_id        prices
count  49672.000000  49672.000000   49672.000000  49672.000000
mean   24850.349775     67.762442      11.728942      9.993282
std    14340.705287     38.315784       5.850779    453.615536
min        1.000000      1.000000       1.000000      1.000000
25%    12432.750000     35.000000       7.000000      4.100000
50%    24850.500000     69.000000      13.000000      7.100000
75%    37268.250000    100.000000      17.000000     11.100000
max    49688.000000    134.000000      21.000000  99999.000000

In [48]:
#Checking the frequency of the prices column
df_prods_clean_no_dups['prices'].value_counts()

prices
2.5        470
5.3        458
6.2        451
2.6        447
5.4        444
          ... 
15.6         1
21.0         1
99999.0      1
14900.0      1
18.3         1
Name: count, Length: 242, dtype: int64

In [50]:
#Removing the items priced at 14900 and 99999
df_prods_clean_no_dups_2 = df_prods_clean_no_dups.drop(df_prods_clean_no_dups[df_prods_clean_no_dups.prices > 14899].index)

In [52]:
#Checking statistics of df_prods_clean_no_dups_2
df_prods_clean_no_dups_2.describe()

product_id      aisle_id  department_id        prices
count  49670.000000  49670.000000   49670.000000  49670.000000
mean   24850.238716     67.761305      11.728770      7.680437
std    14340.931850     38.316060       5.850834      4.199381
min        1.000000      1.000000       1.000000      1.000000
25%    12432.250000     35.000000       7.000000      4.100000
50%    24850.500000     69.000000      13.000000      7.100000
75%    37268.750000    100.000000      17.000000     11.100000
max    49688.000000    134.000000      21.000000     25.000000

In [54]:
#Checking the shape of df_prods_clean_no_dups_2
df_prods_clean_no_dups_2.shape

(49670, 5)

In [56]:
#Exporting df_prods_clean_no_dups_2 dataframe
df_prods_clean_no_dups_2.to_csv(os.path.join(path, '02 Data','Prepared Data', 'products_checked.csv'))

#### 4. Task

In [43]:
#Checking statistics of df_ords
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_week  order_hour_of_day  days_since_prior_order  
count        3.421083e+06       3.421083e+06            3.214874e+06  
mean         2.776219e+00       1.345202e+01            1.111484e+01  
std          2.046829e+00       4.226088e+00            9.206737e+00  
min          0.000000e+00       0.000000e+00            0.000000e+00  
25%          1.000000e+00       1.000000e+01            4.000000e+00  
50%          3.000000e+00       1.300000e+01            7.000000e+00  
75%          5.000000e+00       1.600000e+01            1.500000e+01  
max          6.000000e+00       2.300000e+01            3.000000e+01

Question 2. The data in each column looks fine and all the values make sense, except for the count for the days_since_prior_order column which is less than the count in all the other columns. This indicates that the days_since_prior_order column may be missing some data.

In [45]:
#Checking for mixed-type data in the df_ords dataframe
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

C:\Users\Priya\AppData\Local\Temp\ipykernel_8668\2519883293.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
C:\Users\Priya\AppData\Local\Temp\ipykernel_8668\2519883293.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
C:\Users\Priya\AppData\Local\Temp\ipykernel_8668\2519883293.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
C:\Users\Priya\AppData\Local\Temp\ipykernel_8668\2519883293.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
C:\Users\Priya\AppData\Local\Tem

In [49]:
#Checking the data types of the columns in df_ords
df_ords.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3421083 entries, 0 to 3421082
Data columns (total 8 columns):
 #   Column                  Dtype  
---  ------                  -----  
 0   Unnamed: 0              int64  
 1   order_id                int64  
 2   user_id                 int64  
 3   eval_set                object 
 4   order_number            int64  
 5   orders_day_of_week      int64  
 6   order_hour_of_day       int64  
 7   days_since_prior_order  float64
dtypes: float64(1), int64(6), object(1)
memory usage: 208.8+ MB


In [51]:
#There is no mixed-type data in the df_ords dataframe.

In [53]:
#Checking for missing values in the df_ords dataframe
df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
eval_set                       0
order_number                   0
orders_day_of_week             0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

Question 5. There are only missing values in one column which is the days_since_prior_order column which is missing 206209 values. These values may be missing because there would not be any count for the days since prior order on a customer's first order.

In [63]:
#Address the missing values. First step is to create a subset of the df_ords dataframe that contains the missing values
df_ords_nan = df_ords[df_ords['days_since_prior_order'].isnull() == True]

In [65]:
#Viewing df_ords_nan
df_ords_nan

Unnamed: 0  order_id  user_id eval_set  order_number  \
0                 0   2539329        1    prior             1   
11               11   2168274        2    prior             1   
26               26   1374495        3    prior             1   
39               39   3343014        4    prior             1   
45               45   2717275        5    prior             1   
...             ...       ...      ...      ...           ...   
3420930     3420930    969311   206205    prior             1   
3420934     3420934   3189322   206206    prior             1   
3421002     3421002   2166133   206207    prior             1   
3421019     3421019   2227043   206208    prior             1   
3421069     3421069   3154581   206209    prior             1   

         orders_day_of_week  order_hour_of_day  days_since_prior_order  
0                         2                  8                     NaN  
11                        2                 11                     NaN  
26                        1                 14                     NaN  
39                        6                 11                     NaN  
45                        3                 12                     NaN  
...                     ...                ...                     ...  
3420930                   4                 12                     NaN  
3420934                   3                 18                     NaN  
3421002                   6                 19                     NaN  
3421019                   1                 15                     NaN  
3421069                   3                 11                     NaN  

[206209 rows x 8 columns]

In [69]:
#Checking median of days_since_prior_order column in df_ords
df_ords['days_since_prior_order'].median()

7.0

In [73]:
#Address the missing values. Next step is to replace the missing values with the median.
df_ords.fillna({'days_since_prior_order': 7}, inplace=True)

In [75]:
#Checking if the imputation worked
df_ords[df_ords['order_number']==1]

Unnamed: 0  order_id  user_id eval_set  order_number  \
0                 0   2539329        1    prior             1   
11               11   2168274        2    prior             1   
26               26   1374495        3    prior             1   
39               39   3343014        4    prior             1   
45               45   2717275        5    prior             1   
...             ...       ...      ...      ...           ...   
3420930     3420930    969311   206205    prior             1   
3420934     3420934   3189322   206206    prior             1   
3421002     3421002   2166133   206207    prior             1   
3421019     3421019   2227043   206208    prior             1   
3421069     3421069   3154581   206209    prior             1   

         orders_day_of_week  order_hour_of_day  days_since_prior_order  
0                         2                  8                     7.0  
11                        2                 11                     7.0  
26                        1                 14                     7.0  
39                        6                 11                     7.0  
45                        3                 12                     7.0  
...                     ...                ...                     ...  
3420930                   4                 12                     7.0  
3420934                   3                 18                     7.0  
3421002                   6                 19                     7.0  
3421019                   1                 15                     7.0  
3421069                   3                 11                     7.0  

[206209 rows x 8 columns]

Question 6. I replaced the missing values with the median because it was the best option for dealing with the missing values. There are too many missing values to remove or filter out the data from the dataframe and creating a new variable did not seem to be as good of an option as imputing the median. I chose to impute the median over the mean as the mean seemed to have been influenced by some extreme values.

In [77]:
#Checking for duplicates in the df_ords dataframe
df_ords_dups= df_ords[df_ords.duplicated()]

In [79]:
#Checking df_ords_dups
df_ords_dups

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, eval_set, order_number, orders_day_of_week, order_hour_of_day, days_since_prior_order]
Index: []

Question 7. There are no duplicates in the df_ords dataframe.

Question 8. Since there are no duplicates in the data, there is no need to execute any code to address them.

In [81]:
#Exporting df_ords dataframe
df_ords.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_checked.csv'))